File: api_gateway/websocket_manager.py
"""
WebSocket Manager for Real-time Event Streaming
Handles WebSocket connections, broadcasting, and client management
"""

import asyncio
import json
import logging
import time
from dataclasses import dataclass, asdict
from typing import Dict, List, Optional, Any, Set
from datetime import datetime, timezone
import websockets
from websockets.exceptions import ConnectionClosed, WebSocketException

logger = logging.getLogger(__name__)

@dataclass
class WebSocketClient:
    """Represents a connected WebSocket client"""
    client_id: str
    websocket: Any
    connected_at: datetime
    last_ping: datetime
    filters: Dict[str, Any]
    
    def __post_init__(self):
        if isinstance(self.connected_at, str):
            self.connected_at = datetime.fromisoformat(self.connected_at.replace('Z', '+00:00'))
        if isinstance(self.last_ping, str):
            self.last_ping = datetime.fromisoformat(self.last_ping.replace('Z', '+00:00'))

@dataclass
class EventMessage:
    """Represents an event message to be sent via WebSocket"""
    event_type: str
    data: Dict[str, Any]
    timestamp: datetime
    source: str
    priority: str = "normal"
    
    def __post_init__(self):
        if isinstance(self.timestamp, str):
            self.timestamp = datetime.fromisoformat(self.timestamp.replace('Z', '+00:00'))
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            "event_type": self.event_type,
            "data": self.data,
            "timestamp": self.timestamp.isoformat(),
            "source": self.source,
            "priority": self.priority
        }

class WebSocketManager:
    """
    Manages WebSocket connections and real-time event broadcasting
    
    Features:
    - Client connection lifecycle management
    - Real-time event broadcasting with filtering
    - Health monitoring and metrics
    - Background tasks for metrics and health checks
    """
    
    def __init__(self):
        self.clients: Dict[str, WebSocketClient] = {}
        self.event_history: List[EventMessage] = []
        self.metrics = {
            "total_connections": 0,
            "active_connections": 0,
            "events_sent": 0,
            "last_activity": datetime.now(timezone.utc)
        }
        self.max_history_size = 1000
        self.max_clients = 100
        self._background_tasks: Set[asyncio.Task] = set()
        logger.info("WebSocketManager initialized")
    
    async def connect_client(self, client_id: str, websocket: Any, filters: Dict[str, Any] = None) -> bool:
        """
        Connect a new WebSocket client
        
        Args:
            client_id: Unique identifier for the client
            websocket: WebSocket connection object
            filters: Optional filters for event filtering
            
        Returns:
            bool: True if connected successfully, False otherwise
        """
        if len(self.clients) >= self.max_clients:
            logger.warning(f"Maximum clients ({self.max_clients}) reached, rejecting connection")
            return False
        
        if client_id in self.clients:
            logger.warning(f"Client {client_id} already connected, replacing connection")
            await self.disconnect_client(client_id)
        
        client = WebSocketClient(
            client_id=client_id,
            websocket=websocket,
            connected_at=datetime.now(timezone.utc),
            last_ping=datetime.now(timezone.utc),
            filters=filters or {}
        )
        
        self.clients[client_id] = client
        self.metrics["total_connections"] += 1
        self.metrics["active_connections"] = len(self.clients)
        self.metrics["last_activity"] = datetime.now(timezone.utc)
        
        logger.info(f"Client {client_id} connected, total active: {len(self.clients)}")
        
        # Start background tasks if this is the first client
        if len(self.clients) == 1:
            await self._start_background_tasks()
        
        return True
    
    async def disconnect_client(self, client_id: str) -> bool:
        """
        Disconnect a WebSocket client
        
        Args:
            client_id: Unique identifier for the client
            
        Returns:
            bool: True if disconnected successfully, False if client not found
        """
        if client_id not in self.clients:
            logger.warning(f"Client {client_id} not found for disconnection")
            return False
        
        client = self.clients[client_id]
        
        try:
            if hasattr(client.websocket, 'close'):
                await client.websocket.close()
        except Exception as e:
            logger.error(f"Error closing WebSocket for client {client_id}: {e}")
        
        del self.clients[client_id]
        self.metrics["active_connections"] = len(self.clients)
        self.metrics["last_activity"] = datetime.now(timezone.utc)
        
        logger.info(f"Client {client_id} disconnected, total active: {len(self.clients)}")
        
        # Stop background tasks if no clients remain
        if len(self.clients) == 0:
            await self._stop_background_tasks()
        
        return True
    
    async def broadcast_event(self, event: EventMessage) -> int:
        """
        Broadcast an event to all connected clients
        
        Args:
            event: Event message to broadcast
            
        Returns:
            int: Number of clients that received the event
        """
        if not self.clients:
            return 0
        
        # Add to history
        self.event_history.append(event)
        if len(self.event_history) > self.max_history_size:
            self.event_history.pop(0)
        
        # Prepare message
        message = json.dumps(event.to_dict())
        successful_sends = 0
        clients_to_remove = []
        
        # Send to all connected clients
        for client_id, client in self.clients.items():
            try:
                # Apply client filters if any
                if self._should_send_to_client(event, client):
                    if hasattr(client.websocket, 'send'):
                        await client.websocket.send(message)
                    successful_sends += 1
                    
            except ConnectionClosed:
                logger.info(f"Client {client_id} connection closed during broadcast")
                clients_to_remove.append(client_id)
            except WebSocketException as e:
                logger.error(f"WebSocket error for client {client_id}: {e}")
                clients_to_remove.append(client_id)
            except Exception as e:
                logger.error(f"Unexpected error broadcasting to client {client_id}: {e}")
                clients_to_remove.append(client_id)
        
        # Remove disconnected clients
        for client_id in clients_to_remove:
            await self.disconnect_client(client_id)
        
        self.metrics["events_sent"] += successful_sends
        self.metrics["last_activity"] = datetime.now(timezone.utc)
        
        if successful_sends > 0:
            logger.debug(f"Broadcast event to {successful_sends} clients")
        
        return successful_sends
    
    def _should_send_to_client(self, event: EventMessage, client: WebSocketClient) -> bool:
        """
        Check if an event should be sent to a specific client based on filters
        
        Args:
            event: Event message to check
            client: Client to check against
            
        Returns:
            bool: True if event should be sent to client
        """
        if not client.filters:
            return True
        
        # Check event type filter
        if "event_types" in client.filters:
            if event.event_type not in client.filters["event_types"]:
                return False
        
        # Check source filter
        if "sources" in client.filters:
            if event.source not in client.filters["sources"]:
                return False
        
        # Check priority filter
        if "priority" in client.filters:
            if event.priority not in client.filters["priority"]:
                return False
        
        return True
    
    async def update_client_filters(self, client_id: str, filters: Dict[str, Any]) -> bool:
        """
        Update filters for a specific client
        
        Args:
            client_id: Unique identifier for the client
            filters: New filters to apply
            
        Returns:
            bool: True if updated successfully, False if client not found
        """
        if client_id not in self.clients:
            return False
        
        self.clients[client_id].filters = filters
        logger.info(f"Updated filters for client {client_id}")
        return True
    
    async def ping_client(self, client_id: str) -> bool:
        """
        Send a ping to a specific client
        
        Args:
            client_id: Unique identifier for the client
            
        Returns:
            bool: True if ping sent successfully, False otherwise
        """
        if client_id not in self.clients:
            return False
        
        client = self.clients[client_id]
        
        try:
            if hasattr(client.websocket, 'ping'):
                await client.websocket.ping()
            client.last_ping = datetime.now(timezone.utc)
            return True
        except Exception as e:
            logger.error(f"Error pinging client {client_id}: {e}")
            await self.disconnect_client(client_id)
            return False
    
    async def get_client_info(self, client_id: str) -> Optional[Dict[str, Any]]:
        """
        Get information about a specific client
        
        Args:
            client_id: Unique identifier for the client
            
        Returns:
            Optional[Dict]: Client information or None if not found
        """
        if client_id not in self.clients:
            return None
        
        client = self.clients[client_id]
        return {
            "client_id": client.client_id,
            "connected_at": client.connected_at.isoformat(),
            "last_ping": client.last_ping.isoformat(),
            "filters": client.filters
        }
    
    def get_metrics(self) -> Dict[str, Any]:
        """
        Get current WebSocket metrics
        
        Returns:
            Dict: Current metrics
        """
        return {
            **self.metrics,
            "last_activity": self.metrics["last_activity"].isoformat(),
            "event_history_size": len(self.event_history)
        }
    
    def get_event_history(self, limit: int = 100) -> List[Dict[str, Any]]:
        """
        Get recent event history
        
        Args:
            limit: Maximum number of events to return
            
        Returns:
            List[Dict]: Recent events
        """
        return [event.to_dict() for event in self.event_history[-limit:]]
    
    async def _start_background_tasks(self):
        """Start background tasks for metrics and health checks"""
        # Metrics broadcasting task (every 10 seconds)
        metrics_task = asyncio.create_task(self._metrics_broadcast_task())
        self._background_tasks.add(metrics_task)
        
        # Health check task (every 30 seconds)
        health_task = asyncio.create_task(self._health_check_task())
        self._background_tasks.add(health_task)
        
        logger.info("Background tasks started")
    
    async def _stop_background_tasks(self):
        """Stop all background tasks"""
        for task in self._background_tasks:
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                pass
        
        self._background_tasks.clear()
        logger.info("Background tasks stopped")
    
    async def _metrics_broadcast_task(self):
        """Background task to broadcast metrics every 10 seconds"""
        while True:
            try:
                if self.clients:
                    metrics_event = EventMessage(
                        event_type="metrics",
                        data=self.get_metrics(),
                        timestamp=datetime.now(timezone.utc),
                        source="websocket_manager",
                        priority="low"
                    )
                    await self.broadcast_event(metrics_event)
                
                await asyncio.sleep(10)
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in metrics broadcast task: {e}")
                await asyncio.sleep(10)
    
    async def _health_check_task(self):
        """Background task to check client health every 30 seconds"""
        while True:
            try:
                if self.clients:
                    disconnected_clients = []
                    for client_id in list(self.clients.keys()):
                        success = await self.ping_client(client_id)
                        if not success:
                            disconnected_clients.append(client_id)
                    
                    if disconnected_clients:
                        logger.info(f"Health check removed {len(disconnected_clients)} inactive clients")
                
                await asyncio.sleep(30)
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in health check task: {e}")
                await asyncio.sleep(30)
    
    async def cleanup(self):
        """Clean up resources and close all connections"""
        logger.info("Cleaning up WebSocketManager")
        
        # Stop background tasks
        await self._stop_background_tasks()
        
        # Disconnect all clients
        for client_id in list(self.clients.keys()):
            await self.disconnect_client(client_id)
        
        # Clear history
        self.event_history.clear()
        
        logger.info("WebSocketManager cleanup completed")

# Global instance
_websocket_manager = None

def get_websocket_manager() -> WebSocketManager:
    """Get the global WebSocket manager instance"""
    global _websocket_manager
    if _websocket_manager is None:
        _websocket_manager = WebSocketManager()
    return _websocket_manager

async def create_websocket_manager() -> WebSocketManager:
    """Create a new WebSocket manager instance"""
    return WebSocketManager() 

File: ui/nextjs/src/lib/websocket-manager.ts
/**
 * WebSocket Manager for Next.js
 * Handles WebSocket connections, event broadcasting, and client management
 * Migrated from legacy Python WebSocket manager
 */

import { WebSocket, WebSocketServer } from 'ws'
import { EventEmitter } from 'events'
import { v4 as uuidv4 } from 'uuid'

// Types
export interface WebSocketClient {
  id: string
  websocket: WebSocket
  connectedAt: Date
  lastPing: Date
  filters: Record<string, any>
  tenantId?: string
  userId?: string
  metadata: Record<string, any>
}

export interface EventMessage {
  id: string
  eventType: string
  data: any
  timestamp: string
  source: string
  priority: 'low' | 'normal' | 'high' | 'critical'
  tenantId?: string
  userId?: string
  correlationId?: string
}

export interface WebSocketMetrics {
  totalConnections: number
  activeConnections: number
  eventsSent: number
  lastActivity: Date
  eventHistorySize: number
  clientCount: number
}

export interface WebSocketFilters {
  eventTypes?: string[]
  sources?: string[]
  priority?: string[]
  tenantId?: string
  userId?: string
  search?: string
  timeRange?: string
}

export class WebSocketManager extends EventEmitter {
  private wss: WebSocketServer | null = null
  private clients: Map<string, WebSocketClient> = new Map()
  private eventHistory: EventMessage[] = []
  private metrics: WebSocketMetrics
  private maxHistorySize: number = 1000
  private maxClients: number = 100
  private pingInterval: NodeJS.Timeout | null = null
  private cleanupInterval: NodeJS.Timeout | null = null

  constructor() {
    super()
    this.metrics = {
      totalConnections: 0,
      activeConnections: 0,
      eventsSent: 0,
      lastActivity: new Date(),
      eventHistorySize: 0,
      clientCount: 0
    }
    
    this.startCleanupInterval()
    console.log('WebSocket Manager initialized')
  }

  /**
   * Initialize WebSocket server
   */
  public initialize(server: any): void {
    if (this.wss) {
      console.warn('WebSocket server already initialized')
      return
    }

    this.wss = new WebSocketServer({ server })
    
    this.wss.on('connection', (ws: WebSocket, request: any) => {
      this.handleConnection(ws, request)
    })

    this.wss.on('error', (error: Error) => {
      console.error('WebSocket server error:', error)
      this.emit('error', error)
    })

    console.log('WebSocket server initialized')
  }

  /**
   * Handle new WebSocket connection
   */
  private handleConnection(ws: WebSocket, request: any): void {
    const clientId = uuidv4()
    const url = new URL(request.url, 'http://localhost')
    
    // Extract query parameters
    const tenantId = url.searchParams.get('tenant_id')
    const userId = url.searchParams.get('user_id')
    const eventTypes = url.searchParams.get('event_types')?.split(',') || []
    
    // Create client
    const client: WebSocketClient = {
      id: clientId,
      websocket: ws,
      connectedAt: new Date(),
      lastPing: new Date(),
      filters: {
        eventTypes,
        tenantId,
        userId
      },
      tenantId: tenantId || undefined,
      userId: userId || undefined,
      metadata: {
        userAgent: request.headers['user-agent'],
        ip: request.socket.remoteAddress,
        url: request.url
      }
    }

    // Store client
    this.clients.set(clientId, client)
    this.metrics.totalConnections++
    this.metrics.activeConnections++
    this.metrics.clientCount = this.clients.size
    this.metrics.lastActivity = new Date()

    // Send welcome message
    const welcomeEvent: EventMessage = {
      id: uuidv4(),
      eventType: 'connection',
      data: {
        message: `Connected to WebSocket server as ${clientId}`,
        clientId,
        timestamp: new Date().toISOString()
      },
      timestamp: new Date().toISOString(),
      source: 'websocket_manager',
      priority: 'normal',
      tenantId,
      userId
    }

    this.sendToClient(clientId, welcomeEvent)

    // Set up event handlers
    ws.on('message', (data: Buffer) => {
      this.handleClientMessage(clientId, data.toString())
    })

    ws.on('close', () => {
      this.handleClientDisconnect(clientId)
    })

    ws.on('error', (error: Error) => {
      console.error(`WebSocket error for client ${clientId}:`, error)
      this.handleClientDisconnect(clientId)
    })

    // Emit connection event
    this.emit('client_connected', client)
    console.log(`WebSocket client ${clientId} connected (${this.metrics.activeConnections} active)`)
  }

  /**
   * Handle client message
   */
  private handleClientMessage(clientId: string, message: string): void {
    try {
      const data = JSON.parse(message)
      const client = this.clients.get(clientId)
      
      if (!client) return

      // Update last ping
      client.lastPing = new Date()
      this.metrics.lastActivity = new Date()

      // Handle different message types
      switch (data.type) {
        case 'filters':
          this.updateClientFilters(clientId, data.filters || {})
          break
        case 'ping':
          this.sendToClient(clientId, {
            id: uuidv4(),
            eventType: 'pong',
            data: { timestamp: new Date().toISOString() },
            timestamp: new Date().toISOString(),
            source: 'websocket_manager',
            priority: 'normal'
          })
          break
        case 'subscribe':
          this.handleSubscription(clientId, data)
          break
        case 'unsubscribe':
          this.handleUnsubscription(clientId, data)
          break
        default:
          console.log(`Unknown message type from client ${clientId}:`, data.type)
      }
    } catch (error) {
      console.error(`Error parsing message from client ${clientId}:`, error)
    }
  }

  /**
   * Update client filters
   */
  private updateClientFilters(clientId: string, filters: WebSocketFilters): void {
    const client = this.clients.get(clientId)
    if (!client) return

    client.filters = { ...client.filters, ...filters }
    console.log(`Updated filters for client ${clientId}:`, filters)
    
    this.emit('filters_updated', { clientId, filters })
  }

  /**
   * Handle subscription request
   */
  private handleSubscription(clientId: string, data: any): void {
    const client = this.clients.get(clientId)
    if (!client) return

    const { channel, eventTypes } = data
    
    if (channel) {
      client.filters.channel = channel
    }
    
    if (eventTypes) {
      client.filters.eventTypes = Array.isArray(eventTypes) ? eventTypes : [eventTypes]
    }

    console.log(`Client ${clientId} subscribed to channel: ${channel}, events: ${eventTypes}`)
    this.emit('subscription_updated', { clientId, channel, eventTypes })
  }

  /**
   * Handle unsubscription request
   */
  private handleUnsubscription(clientId: string, data: any): void {
    const client = this.clients.get(clientId)
    if (!client) return

    const { channel, eventTypes } = data
    
    if (channel) {
      delete client.filters.channel
    }
    
    if (eventTypes) {
      client.filters.eventTypes = client.filters.eventTypes?.filter(
        (type: string) => !eventTypes.includes(type)
      )
    }

    console.log(`Client ${clientId} unsubscribed from channel: ${channel}, events: ${eventTypes}`)
    this.emit('subscription_updated', { clientId, channel, eventTypes })
  }

  /**
   * Handle client disconnect
   */
  private handleClientDisconnect(clientId: string): void {
    const client = this.clients.get(clientId)
    if (!client) return

    // Clean up client
    this.clients.delete(clientId)
    this.metrics.activeConnections--
    this.metrics.clientCount = this.clients.size
    this.metrics.lastActivity = new Date()

    // Emit disconnect event
    this.emit('client_disconnected', client)
    console.log(`WebSocket client ${clientId} disconnected (${this.metrics.activeConnections} active)`)
  }

  /**
   * Send message to specific client
   */
  public sendToClient(clientId: string, event: EventMessage): boolean {
    const client = this.clients.get(clientId)
    if (!client || client.websocket.readyState !== WebSocket.OPEN) {
      return false
    }

    try {
      client.websocket.send(JSON.stringify(event))
      this.metrics.eventsSent++
      this.metrics.lastActivity = new Date()
      return true
    } catch (error) {
      console.error(`Error sending message to client ${clientId}:`, error)
      return false
    }
  }

  /**
   * Broadcast event to all connected clients
   */
  public broadcastEvent(event: EventMessage): number {
    let successfulSends = 0
    const clientsToRemove: string[] = []

    // Add to history
    this.eventHistory.push(event)
    if (this.eventHistory.length > this.maxHistorySize) {
      this.eventHistory.shift()
    }
    this.metrics.eventHistorySize = this.eventHistory.length

    // Send to all connected clients
    for (const [clientId, client] of this.clients) {
      try {
        // Apply client filters
        if (this.shouldSendToClient(event, client)) {
          if (this.sendToClient(clientId, event)) {
            successfulSends++
          }
        }
      } catch (error) {
        console.error(`Error broadcasting to client ${clientId}:`, error)
        clientsToRemove.push(clientId)
      }
    }

    // Remove disconnected clients
    clientsToRemove.forEach(clientId => {
      this.handleClientDisconnect(clientId)
    })

    if (successfulSends > 0) {
      console.debug(`Broadcast event to ${successfulSends} clients`)
    }

    return successfulSends
  }

  /**
   * Check if event should be sent to client based on filters
   */
  private shouldSendToClient(event: EventMessage, client: WebSocketClient): boolean {
    const { filters } = client

    // Check event type filter
    if (filters.eventTypes && filters.eventTypes.length > 0) {
      if (!filters.eventTypes.includes(event.eventType)) {
        return false
      }
    }

    // Check source filter
    if (filters.sources && filters.sources.length > 0) {
      if (!filters.sources.includes(event.source)) {
        return false
      }
    }

    // Check priority filter
    if (filters.priority && filters.priority.length > 0) {
      if (!filters.priority.includes(event.priority)) {
        return false
      }
    }

    // Check tenant isolation
    if (filters.tenantId && event.tenantId) {
      if (filters.tenantId !== event.tenantId) {
        return false
      }
    }

    // Check user isolation
    if (filters.userId && event.userId) {
      if (filters.userId !== event.userId) {
        return false
      }
    }

    return true
  }

  /**
   * Get client by ID
   */
  public getClient(clientId: string): WebSocketClient | undefined {
    return this.clients.get(clientId)
  }

  /**
   * Get all clients
   */
  public getAllClients(): WebSocketClient[] {
    return Array.from(this.clients.values())
  }

  /**
   * Get clients by tenant
   */
  public getClientsByTenant(tenantId: string): WebSocketClient[] {
    return Array.from(this.clients.values()).filter(client => client.tenantId === tenantId)
  }

  /**
   * Get clients by user
   */
  public getClientsByUser(userId: string): WebSocketClient[] {
    return Array.from(this.clients.values()).filter(client => client.userId === userId)
  }

  /**
   * Get metrics
   */
  public getMetrics(): WebSocketMetrics {
    return { ...this.metrics }
  }

  /**
   * Get event history
   */
  public getEventHistory(limit: number = 100): EventMessage[] {
    return this.eventHistory.slice(-limit)
  }

  /**
   * Start ping interval
   */
  public startPingInterval(intervalMs: number = 30000): void {
    if (this.pingInterval) {
      clearInterval(this.pingInterval)
    }

    this.pingInterval = setInterval(() => {
      const pingEvent: EventMessage = {
        id: uuidv4(),
        eventType: 'ping',
        data: { timestamp: new Date().toISOString() },
        timestamp: new Date().toISOString(),
        source: 'websocket_manager',
        priority: 'normal'
      }

      this.broadcastEvent(pingEvent)
    }, intervalMs)

    console.log(`Started ping interval: ${intervalMs}ms`)
  }

  /**
   * Stop ping interval
   */
  public stopPingInterval(): void {
    if (this.pingInterval) {
      clearInterval(this.pingInterval)
      this.pingInterval = null
      console.log('Stopped ping interval')
    }
  }

  /**
   * Start cleanup interval
   */
  private startCleanupInterval(): void {
    this.cleanupInterval = setInterval(() => {
      this.cleanupInactiveClients()
    }, 60000) // Run every minute
  }

  /**
   * Clean up inactive clients
   */
  private cleanupInactiveClients(): void {
    const now = new Date()
    const inactiveThreshold = 5 * 60 * 1000 // 5 minutes

    for (const [clientId, client] of this.clients) {
      const timeSinceLastPing = now.getTime() - client.lastPing.getTime()
      
      if (timeSinceLastPing > inactiveThreshold) {
        console.log(`Cleaning up inactive client ${clientId}`)
        this.handleClientDisconnect(clientId)
      }
    }
  }

  /**
   * Shutdown WebSocket server
   */
  public shutdown(): void {
    // Stop intervals
    this.stopPingInterval()
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval)
      this.cleanupInterval = null
    }

    // Close all client connections
    for (const [clientId, client] of this.clients) {
      try {
        client.websocket.close()
      } catch (error) {
        console.error(`Error closing client ${clientId}:`, error)
      }
    }

    // Close server
    if (this.wss) {
      this.wss.close()
      this.wss = null
    }

    // Clear clients
    this.clients.clear()
    this.metrics.activeConnections = 0
    this.metrics.clientCount = 0

    console.log('WebSocket server shutdown complete')
  }
}

// Singleton instance
let websocketManager: WebSocketManager | null = null

export function getWebSocketManager(): WebSocketManager {
  if (!websocketManager) {
    websocketManager = new WebSocketManager()
  }
  return websocketManager
}

export function shutdownWebSocketManager(): void {
  if (websocketManager) {
    websocketManager.shutdown()
    websocketManager = null
  }
}
